sdk/lib/atl/atlimage.h

```diff
@@ -93,28 +93,40 @@ class CImage
 
     HDC GetDC() const throw()
     {
-        if (m_hDC)
-            return m_hDC;
+        ATLASSERT(m_nDCRefCount >= 0);
+
+        if (::InterlockedIncrement(&m_nDCRefCount) == 1)
+        {
+            ATLASSERT(m_hDC == NULL);
+            ATLASSERT(m_hOldBitmap == NULL);
+            m_hDC = ::CreateCompatibleDC(NULL);
+            m_hOldBitmap = (HBITMAP)::SelectObject(m_hDC, m_hBitmap);
+        }
 
-        m_hDC = ::CreateCompatibleDC(NULL);
-        m_hOldBitmap = (HBITMAP)::SelectObject(m_hDC, m_hBitmap);
         return m_hDC;
     }
 
     void ReleaseDC() const throw()
     {
-        ATLASSERT(m_hDC);
+        ATLASSERT(m_nDCRefCount > 0);
 
-        if (m_hDC == NULL)
+        if (::InterlockedDecrement(&m_nDCRefCount) > 0)
             return;
 
+        ATLASSERT(m_hDC != NULL);
+        ATLASSERT(m_hOldBitmap != NULL);
+
         if (m_hOldBitmap)
         {
             ::SelectObject(m_hDC, m_hOldBitmap);
             m_hOldBitmap = NULL;
         }
-        ::DeleteDC(m_hDC);
-        m_hDC = NULL;
+
+        if (m_hDC)
+        {
+            ::DeleteDC(m_hDC);
+            m_hDC = NULL;
+        }
```

Comment:
Taking your previous considerations into account it seems that the following pseudo-code would be simpler:
```c++
if (!m_hDC)
{
    ASSERT(!m_hOldBitmap);
    return;
}

if (m_hOldBitmap)
{
// Do as in the original code
}

::DeleteDC(m_hDC);
m_hDC = NULL;
```

Reply:
Fixed.